ext/openssl/openssl.c

```diff
@@ -4228,180 +4228,280 @@ static EVP_PKEY *php_openssl_pkey_init_dh(zval *data, bool *is_private)
 static bool php_openssl_pkey_init_legacy_ec(EC_KEY *eckey, zval *data, bool *is_private) {
+	BIGNUM *p = NULL, *a = NULL, *b = NULL, *g_x = NULL, *g_y = NULL, *order = NULL, *cofactor = NULL;
+	BIGNUM *x = NULL, *y = NULL, *d = NULL;
 	EC_GROUP *group = NULL;
-	EC_POINT *pnt = NULL;
-	BIGNUM *d = NULL;
-	zval *bn;
-	zval *x;
-	zval *y;
+	EC_POINT *point_g = NULL;
+	EC_POINT *point_q = NULL;
+	BN_CTX *bctx = BN_CTX_new();
+	zval *curve_name_zv = zend_hash_str_find(Z_ARRVAL_P(data), "curve_name", sizeof("curve_name") - 1);
+	zval *generator_zv = zend_hash_str_find(Z_ARRVAL_P(data), "generator", sizeof("generator") - 1);
+	zval *seed_zv = zend_hash_str_find(Z_ARRVAL_P(data), "seed", sizeof("seed") - 1);
 
 	*is_private = false;
 
-	if ((bn = zend_hash_str_find(Z_ARRVAL_P(data), "curve_name", sizeof("curve_name") - 1)) != NULL &&
-			Z_TYPE_P(bn) == IS_STRING) {
-		int nid = OBJ_sn2nid(Z_STRVAL_P(bn));
-		if (nid != NID_undef) {
-			group = EC_GROUP_new_by_curve_name(nid);
-			if (!group) {
-				php_openssl_store_errors();
+	if(curve_name_zv && Z_TYPE_P(curve_name_zv) == IS_STRING && Z_STRLEN_P(curve_name_zv) > 0) {
+		int nid = OBJ_sn2nid(Z_STRVAL_P(curve_name_zv));
+		group = EC_GROUP_new_by_curve_name(nid);
+		if (nid == NID_undef || !group) {
+				php_error_docref(NULL, E_WARNING, "Unknown curve name");
 				goto clean_exit;
-			}
-			EC_GROUP_set_asn1_flag(group, OPENSSL_EC_NAMED_CURVE);
-			EC_GROUP_set_point_conversion_form(group, POINT_CONVERSION_UNCOMPRESSED);
-			if (!EC_KEY_set_group(eckey, group)) {
-				php_openssl_store_errors();
+		}
+
+		EC_GROUP_set_asn1_flag(group, OPENSSL_EC_NAMED_CURVE);
+	} else {
+		OPENSSL_PKEY_SET_BN(data, p);
+		OPENSSL_PKEY_SET_BN(data, a);
+		OPENSSL_PKEY_SET_BN(data, b);
+		OPENSSL_PKEY_SET_BN(data, order);
+		OPENSSL_PKEY_SET_BN(data, g_x);
+		OPENSSL_PKEY_SET_BN(data, g_y);
+		OPENSSL_PKEY_SET_BN(data, cofactor);
+
+		if (!(p && a && b && order)) {
+			php_error_docref(NULL, E_WARNING, "Missing params: need \"p\", \"a\", \"b\", \"order\" ");
+			goto clean_exit;
+		}
+
+		if(!(group = EC_GROUP_new_curve_GFp(p, a, b, bctx))) {
+			goto clean_exit;
+		}
+
+		point_g = EC_POINT_new(group);
+
+		if (generator_zv && Z_TYPE_P(generator_zv) == IS_STRING && Z_STRLEN_P(generator_zv) > 0) {
+			EC_POINT_oct2point(group, point_g, (unsigned char *)Z_STRVAL_P(generator_zv), Z_STRLEN_P(generator_zv), bctx);
+		} else if(g_x && g_y) {
+			if (!EC_POINT_set_affine_coordinates(group, point_g, g_x, g_y, bctx)) {
 				goto clean_exit;
 			}
+		} else {
+			php_error_docref(
+				NULL, E_WARNING, "Missing params: need \"generator\" or \"g_x\" and \"g_y\" ");
+			goto clean_exit;
+		}
+
+		if (seed_zv && Z_TYPE_P(seed_zv) == IS_STRING && Z_STRLEN_P(seed_zv) > 0) {
+			EC_GROUP_set_seed(group, (unsigned char *)Z_STRVAL_P(seed_zv), Z_STRLEN_P(seed_zv));
+		}
+
+		if (cofactor) {
+			EC_GROUP_set_generator(group, point_g, order, cofactor);
+		} else {
+			EC_GROUP_set_generator(group, point_g, order, BN_value_one());
```

Comment:
Check for error. Can you explain why using `BN_value_one` makes sense rather than making generator required?

Reply:
Oh! This might be my mistake. The cofactor here can directly pass a NULL value as the default value, so this "if" statement is not necessary.